src/tweak.h

```diff
@@ -219,27 +231,47 @@ class CTweak : public CTweakBase
     virtual std::string GetHelp() const { return help; }
     virtual UniValue Get() const
     {
-        std::lock_guard<std::mutex> lck(cs_tweak);
+        std::lock_guard<std::recursive_mutex> lck(cs_tweak);
         return UniValue(value);
     }
-    virtual UniValue Set(const UniValue &v)
+
+    virtual UniValue Set(const UniValue &v) // Returns NullUnivalue or an error string
     {
-        std::lock_guard<std::mutex> lck(cs_tweak);
+        std::lock_guard<std::recursive_mutex> lck(cs_tweak);
+        DataType prior = value;
         fill(v, value);
-
-        // Returns NullUnivalue or an error string
+        if (eventCb)
+        {
+            std::string result = eventCb(prior, this, false);
+            if (!result.empty())
+                return UniValue(result);
+        }
         return NullUniValue;
     }
 
+    virtual std::string Validate(const UniValue &val)
+    {
+        std::lock_guard<std::recursive_mutex> lck(cs_tweak);
+        if (eventCb)
+        {
+            DataType candidate;
+            fill(val, candidate);
+            std::string result = eventCb(candidate, this, true);
+            if (!result.empty())
+                return result;
+        }
+        return std::string();
+    };
+
     virtual DataType Value() const
     {
-        std::lock_guard<std::mutex> lck(cs_tweak);
+        std::lock_guard<std::recursive_mutex> lck(cs_tweak);
         return value;
     }
 
     CTweak &operator=(const DataType &d)
     {
-        std::lock_guard<std::mutex> lck(cs_tweak);
+        std::lock_guard<std::recursive_mutex> lck(cs_tweak);
         value = d;
```

Comment:
why change to recursive_mutex?  I don't see where we need to take the lock recursively, we're only ever dealing with one tweak at a time when looking up internal data.

Reply:
nvm...i see why you did it.